firedrake/functionspaceimpl.py

```diff
@@ -308,6 +348,54 @@ def boundary_nodes(self, sub_domain):
     def collapse(self):
         return type(self).create(self.topological.collapse(), self.mesh())
 
+    @classmethod
+    def make_function_space(cls, mesh, element, name=None):
+        r"""Factory method for :class:`.WithGeometryBase`."""
+        mesh.init()
+        topology = mesh.topology
+        if isinstance(element, finat.ufl.MixedElement) and not isinstance(element, (finat.ufl.VectorElement, finat.ufl.TensorElement)):
+            spaces = [cls.make_function_space(topology, e) for e in element.sub_elements]
+            new = MixedFunctionSpace(spaces, name=name)
+        else:
+            # Check that any Vector/Tensor/Mixed modifiers are outermost.
+            check_element(element)
+            if element.family() == "Real":
+                new = RealFunctionSpace(topology, element, name=name)
+            else:
+                new = FunctionSpace(topology, element, name=name)
+        if mesh is not topology:
```

Comment:
When would this branch not be followed? i.e. when `mesh is topology`?
If line 367 isn't called and `cls` is a `FiredrakeDualSpace` then the return value of this function is a `FunctionSpace` not a `FiredrakeDualSpace`.

Reply:
There is recursion (of depth 1) in the case of a MixedElement, where we build a list of `impl.FunctionSpaces` (which are neither primal or dual) on `mesh.topology` (as opposed to `mesh`). Only the outermost caller will produce a `WithGeometryBase` instance being either primal (WithGeometry) or dual (FiredrakeDualSpace). Terrible class names!

Reply:
Ok that's good. They are definitely confusingly named!